Integrate Login.gov SETs

elipe17 opened the pull request:
## Summary of Changes
- Implemented API endpoint for Login.gov SET management
- Use Login.gov SETs to handle user email updates async instead of at login

Pull request closes #5268 

## How to Test

- Deploy this PR to the `raft` environment. 
- Login to to the raft app, go to your profile, and go to your login.gov account
- Update your password and verify the backend tracks the event by view the Security Event Tokens in the DAC
- Delete your account in login.gov, verify an Account Purged event is generated for your user, and that your user has the `is_active` flag set to `False`.

<img width="1355" height="442" alt="Screenshot 2025-08-19 at 4 56 53 PM" src="https://github.com/user-attachments/assets/aa75890b-b2c6-41a2-8b5f-75882bdf4a27" />

- Recreated your account in login.gov with the same email
- Login to the `raft` app again and verify you have access to everything you did prior to deleting the login.gov account
- Verify in the Log entries that you see an entry indicating your user had it's `login_gov_uuid` updated

<img width="1330" height="415" alt="Screenshot 2025-08-19 at 4 57 09 PM" src="https://github.com/user-attachments/assets/ffdb1870-2cd9-4326-bb9a-54ad2911dd33" />

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [ ] Login.gov partner portal updated to send SETs to the `raft` app, `develop` app, and `staging` app. (Production updates require a follow-up request to Login.gov after successful dev and staging testing.)  
- [ ] Backend updated to integrate with and store Login.gov SETs.  
- [ ] Nginx config updated to proxy Login.gov SETs.  
- [ ] Testing Checklist has been run and all tests pass.  
- [ ] README updated, if necessary.  
- [ ] [Release digest](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/JSIbkxujKUxhKv2cXFEg/) updated with testing steps & relevant media.  

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Add implementation for handling login.gov SETs
- Added new model and migration
- Added view to support receiving SETs
- Updated settings

elipe17 committed:
- Fixed login flow

elipe17 committed:
- remove unused param and add missing return vals

elipe17 committed:
- remove comment

elipe17 committed:
- Add tests for event handler

elipe17 committed:
- remove unused import

elipe17 committed:
- Update permissions for new model

elipe17 committed:
- remove unused parameter

elipe17 committed:
- Add fix for failing ams test

elipe17 committed:
- linting

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 5268-login-gov-set-integation

elipe17 committed:
- Added text choice field for event type to keep consistent
- Regenerated migration
- Updated tests to use new Enum
- Added tests specifically for login_gov_uuid change

elipe17 committed:
- Update view to throw validation errors
- Add tests for security event view

elipe17 committed:
Merge branch 'develop' into 5268-login-gov-set-integation

elipe17 committed:
- Add DAC logging to help with testing

elipe17 committed:
Merge branch 'develop' into 5268-login-gov-set-integation

elipe17 committed:
- Update admin models for readonly use
- Remove DAC logging from SET view
- Add DAC logging when a user account is recreated

elipe17 committed:
- re-add missing admin model class

elipe17 committed:
- Move log statement before clearing user variable

elipe17 committed:
- Update event handler so that it does NOT change approval status. Only change whether user is active or other meta data
- Update tests to reflect this
- Log the decoded data on uuid update after purge

elipe17 committed:
- Use sub directly

elipe17 committed:
- Update tests to use raw sub

elipe17 committed:
Merge branch 'develop' into 5268-login-gov-set-integation

elipe17 committed:
- Update migration to have all SET types
- Add handlers for all types
- Update handler with email or sub logic
- Remove email update logic from login.py since SETs will now handle it

elipe17 committed:
- Resolve failing auth tests

elipe17 committed:
- Fix name references

elipe17 committed:
- Add tests for new handlers

elipe17 committed:
- rename file and add test for user email change

elipe17 committed:
- Remove login.gov test since it is handled by security events now

elipe17 committed:
Merge branch 'develop' into 5268-login-gov-set-integation

raftmsohani commented on tdrs-backend/tdpservice/security/event_handler.py:
```diff
@@ -0,0 +1,182 @@
+
+    def _get_emails(security_event):
+        """Get the old and new emails from the event data."""
+        event_data = security_event.event_data
```
what happens if security_event is malformed?

elipe17 replied:
From the perspective of exceptions, there is exception handling throughout the SET flow. If an exception occurs during the handling of an event, it is caught and the event is not serialized to the database.

From the data perspective, if the request contains an event (this is validated in the view) then the event will contain event data with a subject as specified by [OpenID RISC Event Types 1.0](https://openid.net/specs/openid-risc-event-types-1_0-ID1.html). The content of the subject is variable but each handler knows what it's specific subject should contain to handle the event. All other aspects of the request are validated before ever reaching this point. So the created model should be valid unless login.gov changes the spec to not match the OpenID spec.

raftmsohani commented on tdrs-backend/tdpservice/security/event_handler.py:
```diff
@@ -0,0 +1,182 @@
+    def handle_event(cls, event_type, event_data, decoded_jwt):
+        """Handle specific event types."""
+        try:
+            subject = event_data.get("subject")
```
my suggestion is to add a schema and always check the API response with the schema. We have had instances that login.gov changed their api, which in that case it will be hard to investigate

jtimpe replied:
agree with this. i also think much of both the view and event handler classes could be represented (with a little finagling for the jwt stuff) as a DRF serializer/modelview and that would simplify things. maybe some future work that could come after we see what issues we have. as it is, it seems to work well

elipe17 replied:
I could see converting this endoint to a viewset with a serializer in the future if we want to expose these events on the frontend or to some other consumer. I don't see how we could kill off the event handler though since we still need to do things with the event after the view/serializer creates the model.

raftmsohani commented on tdrs-backend/tdpservice/security/test/test_views.py:
```diff
@@ -67,3 +71,291 @@ def test_generate_new_token(client):
+            "jti": "test_jti"
+          }
+        """
+        return "eyJhbGciOiJSUzI1NiIsImtpZCI6InRlc3Rfa2lkIn0.eyJldmVudHMiOnsiaHR0cHM6Ly9zY2hlbWFzLm9wZW5pZC5uZXQvc2VjZXZlbnQvcmlzYy9ldmVudC10eXBlL2FjY291bnQtZGlzYWJsZWQiOnsic3ViamVjdCI6eyJzdWIiOiI2N2Y0Y2JmNS0xNjE1LTRhYTQtOWU2My0yZjNiMjcwYWNmMDUifX19LCJpc3MiOiJodHRwczovL2xvZ2luLmdvdiIsImlhdCI6MTYyMDAwMDAwMCwianRpIjoidGVzdF9qdGkifQ.signature"
```
nit: my suggestion is to use ```encoded_jwt = jwt.encode(to_encode, secret_key, algorithm=algorithm)``` to encode a human readable json data

raftmsohani commented on tdrs-backend/tdpservice/users/api/login.py:
```diff
@@ -116,6 +119,60 @@ def verify_email(self, email):
+        User = get_user_model()
+
+        # Check if a user with the same email already exists
+        user = User.objects.filter(email=email).first()
```
technically user can have blank email, and we are setting username as email, so I think you will need to change this to: user = User.objects.filter(username=email).first()

jtimpe commented on tdrs-backend/tdpservice/security/event_handler.py:
```diff
@@ -0,0 +1,182 @@
+
+        # Set login_gov_uuid to null and deactivate user
+        user.login_gov_uuid = None
+        user.is_active = False
```
should we reset the approval status as well? or we're good with a user being "auto-approved" following the purge/recreation of the account? i think this makes sense since they'd been previously approved, just want to confirm

elipe17 replied:
I talked with Alex and she agreed that the pre-existing approval status should not be modified regardless of the event.

jtimpe commented on tdrs-backend/tdpservice/security/event_handler.py:
```diff
@@ -0,0 +1,182 @@
+                except (ValueError, TypeError) as e:
+                    logger.warning(f"Error converting timestamp {iat_timestamp}: {e}")
+
+            security_event = SecurityEventToken.objects.create(
```
are we tracking all events with `SecurityEventToken`s, even when we don't do anything except log in the handler?

elipe17 replied:
I decided to implement everything anyways because it could come in handy in the future for audits or things like that. Or if we ever need to do more complicated things with the events we have them ready to be implemented.

jtimpe commented on tdrs-backend/tdpservice/security/event_handler.py:
```diff
@@ -0,0 +1,182 @@
+        """Handle email-recycled event."""
+        new_email, old_email = SecurityEventHandler._get_emails(security_event)
+        user = security_event.user
+        logger.info(f"User {user.username} recycled extra email address {old_email}")
```
what's recycled? adding and removing email addresses only triggered "Recovery Information Changed" events for me

elipe17 replied:
I believe you need to add extra emails onto your account and then remove them. From login.govs docs.
```
Login.gov pushes this event when a user removes an email address from their account, freeing up the email address as an identifier.
```

elipe17 replied:
Sounds like you did that though?

jtimpe replied:
that is what i did. i tested again and still only get the "Recovery Information Changed." i could be doing it wrong. all good either way, just curious

elipe17 replied:
I also tested that flow and got the "Recovery Information Changed" event. Maybe some wires crossed on login.gov's side with respect to triggers for the events

jtimpe commented on tdrs-backend/tdpservice/users/api/login.py:
```diff
@@ -116,6 +119,60 @@ def verify_email(self, email):
+            logger_context = {
+                "user_id": user.id,
+                "content_type": SecurityEventToken,
+                "object_id": user.id,
```
since the `content_type` is `SecurityEventToken`, i think your `object_id` needs to be a `SecurityEventToken` object as well (or `content_type` has to be `User`). currently nothing shows in the object column in DAC. SET and user both make sense to me

jtimpe reviewed (commented):
Agree with mo's comments and some minor comments/questions of my own. Overall works great though!

elipe17 committed:
- update email to username
- make object id the event token

elipe17 committed:
- Add true RSA jwt encoding for tests
- Add missing return from _get_public_key function

elipe17 committed:
- Add null check for content type
- Update factory so that it actually behaves how users are created in reality

elipe17 committed:
- remove comment
- Update user check to use username only
- Add default for subject which will cause exception when used

elipe17 committed:
Merge branch 'develop' into 5268-login-gov-set-integation

elipe17 committed:
- Update permissions ID

elipe17 commented:
<img width="1640" height="840" alt="Screenshot 2025-08-26 at 11 21 54 AM" src="https://github.com/user-attachments/assets/cd4dcf0e-7c15-489f-b11e-12dbebd63963" />

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 5268-login-gov-set-integation

jtimpe reviewed (approved):
Good work!

raftmsohani committed:
Merge branch 'develop' into 5268-login-gov-set-integation

raftmsohani commented:
@elipe17  you have a conflict otherwise I can go ahead and approve. Already tested this 2 weeks ago

jtimpe committed:
Merge branch 'develop' into 5268-login-gov-set-integation

elipe17 commented:
Update. Not merging this yet because Alex wants to wait for it to go into the next release. She will be out and does not want to push something to prod that may require back and forth from login.gov while she is out.

elipe17 committed:
Merge branch 'develop' into 5268-login-gov-set-integation

elipe17 committed:
Merge branch 'develop' into 5268-login-gov-set-integation

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 5268-login-gov-set-integation

elipe17 committed:
- lint/resolve merge conflict

elipe17 committed:
Merge branch 'develop' into 5268-login-gov-set-integation

elipe17 committed:
Merge branch 'develop' into 5268-login-gov-set-integation
